### ISSUE TITLE
CTGAN/CopulaGAN: Hyperparameter Tuning & Evaluation Metric

### ISSUE THREAD

**Issue by AKAgarwal01** [sdv-dev/SDV#287]
* SDV version:                   0.6.0
* Python version:               3.7.0
* Operating System:          Windows/Linux

### Description
Background: Trying to generate synthetic tabular data using CTGAN/CopulaGAN for a **Multi-Classification Task** (20 possible labels) where my **real training data is in order of 10^5 to 10^7** but is **highly imbalanced** (70% belongs to 5 labels and 30% to 15 labels) and with **90 columns** (input features).

### Suggestions
**1. What to use for while generating a CTGAN/CopulaGAN model?**
    Imbalanced Real Data or Balanced Real Data

**2. Hyperparameter Tuning:**
    Since my training dataset is huge, generation of a model takes long time. Therefore, I want to ask if there is any possible way    to tune the Hyperparameters of a model before I actually start training the model with real data to generate the synthetic one.

The closest I came across is RayTune, Optuna, HyperOpt and BayesianOptimization. Could you comment on the idea of tuning my Hyperparameters with respect to my real dataset with any possible Optimization approaches (mentioned above or otherwise)?

**3. Clarity on Evaluation Metric**
a. In Detection Metrics, [https://sdv.dev/SDV/user_guides/evaluation/single_table_metrics.html](https://sdv.dev/SDV/user_guides/evaluation/single_table_metrics.html), it says LogisticDetection and SVCDetection give 1 minus the average ROC AUC score across all the cross validation splits.

b. However, in Evaluation Framework(How was the obtained score computed?),  [https://sdv.dev/SDV/user_guides/evaluation/evaluation_framework.html](https://sdv.dev/SDV/user_guides/evaluation/evaluation_framework.html),  it says the goal of LogisticDetection and SVCDetection is to MAXIMIZE the score which lies in [0, 1].

So do 3.a and 3.b say the same thing and our scores for LogisticDetection and SVCDetection should be closer to 1 for better quality of generated synthetic data? 

**4. Plotting generator and discriminator losses after each epoch**
    a. Does EPOCH losses give an absolute estimate of optimization?
    b. Considering my classification task (summarized in BACKGROUND) what **ranges of epoch and batch size** I should try on with real data of order 10^5 to 10^7 and number of column in dataset about 90?
    c. Is it possible to plot both the losses?

Thank you for your time!

**Comment by csala** [comment]
Hi @AKAgarwal01. All these are very interesting questions.

I'll try responding to some of your points:

> **1. What to use for while generating a CTGAN/CopulaGAN model?**
> Imbalanced Real Data or Balanced Real Data

That depends on what your goal is. If your goal is to build a replica of the dataset, I would suggest using the real data as it is, imbalanced. On the other side, if your goal is to augment the data in order to balance it, you may benefit from subsampling the oversampled classes, or even completely drop them and model/sample the undersampled ones only.

> **2. Hyperparameter Tuning:**
> Since my training dataset is huge, generation of a model takes long time. Therefore, I want to ask if there is any possible way to tune the Hyperparameters of a model before I actually start training the model with real data to generate the synthetic one.
> 
> The closest I came across is RayTune, Optuna, HyperOpt and BayesianOptimization. Could you comment on the idea of tuning my Hyperparameters with respect to my real dataset with any possible Optimization approaches (mentioned above or otherwise)?

Since there are multiple models to choose from (CTGAN, TVAE, CopulaGAN...), and also multiple strategies (Balanced, Unbalanced...) I would recommend you to combine Bayesian Tuning with Multi-armed Bandit Selection, so you want do both the model selection and the hyperparameter tuning at once. For this, please have a look at [BTB](https://github.com/MLBazaar/BTB), which has also been built by the DAI-Lab team and does a great job at [combining both tuning and selection](https://github.com/MLBazaar/BTB/blob/master/tutorials/03_Session.ipynb)

Also, since the training time is high, you may consider these two options:
- start with a smaller subsample to get a notion of the ideal models and hyperparameter ranges, and then increase the data size for a second round of fine tuning.
- In case of CopulaGAN, since the marginal distribution selection takes some time and should also select the same, I would do a first training with the default hyperparameters to let it select the best marginal distributions and then I would just pin them during the hyperparameter tuning process (so, capture them using `model.get_distributions()` and then pass the given dict as the `field_distributions` argument)

> **3. Clarity on Evaluation Metric**
> a. In Detection Metrics, [https://sdv.dev/SDV/user_guides/evaluation/single_table_metrics.html](url), it says LogisticDetection and SVCDetection give 1 minus the average ROC AUC score across all the cross validation splits.
> 
> b. However, in Evaluation Framework(How was the obtained score computed?), [https://sdv.dev/SDV/user_guides/evaluation /evaluation_framework.html](url), it says the goal of LogisticDetection and SVCDetection is to MAXIMIZE the score which lies in [0, 1].
> 
> So do 3.a and 3.b say the same thing and our scores for LogisticDetection and SVCDetection should be closer to 1 for better quality of generated synthetic data?

Yes, your last sentence is right. These return a score which is always in the [0, 1] range, being 0 the worst possible scenario (really bad data) and 1 the best possible scenario (perfect data). Internally, they do:
- compute a ROC AUC score, which stays between 0.5 (random guess) and 1 (perfect classification)
- slightly normalize the score to make it be between 0 (random guess) and 1 (perfect classification)
- return 1 - the obtained value, making the output be between 0 (perfect classification, meaning very bad data) and 1 (random guess, meaning indistinguishable data)

> **4. Plotting generator and discriminator losses after each epoch**
> a. Does EPOCH losses give an absolute estimate of optimization?

No. In the GAN scenarios the loss of each component (generator and discriminator) only indicate how well that component is doing with regards to the other one, but that does not represent an absolute meaning of how well they are doing. In order to evaluate how well the generator is doing in an absolute way you have is to run the model through external metrics, which right now can be done only after the training has finished.

> b. Considering my classification task (summarized in BACKGROUND) what **ranges of epoch and batch size** I should try on with real data of order 10^5 to 10^7 and number of column in dataset about 90?

This depends too much on the actual data to provide an exact answer. In general, CTGAN benefits from having large batch sizes (as long as you have enough resources) and a large number of epochs.

Overall my recommendation for your case would be to:
- Start with the default epochs value to do model selection and broad hyperparameter tuning
- Take the best selected model and do several trials to see for how many epochs you can train while still getting noticeable improvements.
- Once you know that number of epochs, use that to do fine hyperparameter tuning.
- Once you know the ideal hyperparameters, fit one last time using them and a slightly higher number of epochs.

> c. Is it possible to plot both the losses?

It is not. Right now the losses are only printed to console.

**Comment by Aditya-Agarwal-018** [comment]
Hello,

Thank you for your detailed explanation! 

I just have one more question related to Hyperparameter Tuning procedure using [BTB](https://github.com/MLBazaar/BTB):

===========================CODE for HP Tuning===================================
### Define all the models to be considered
models = {
    'CTGAN': CTGAN,
    'CopulaGAN': CopulaGAN
}

### Define scoring function
def scoring_function(model_name, hyperparameter_values):
    model_class = models[model_name]
    model_instance = model_class(**hyperparameter_values)
    **scores = evaluate(synthetic_data_input, real_data_global)**                               # **Doubt in this line**
    print(model_name,': ',scores)
    return scores

### Defining the tunables
tunables = {
    'CTGAN': Tunable({
        'batch_size': hp.IntHyperParam(min=100, max=1000,default=500, step=100),
        'embedding_dim':hp.IntHyperParam(min=64, max=512,default=128, step=64),
    }),
    'CopulaGAN': Tunable({
        'batch_size': hp.IntHyperParam(min=100, max=1000,default=500, step=100),
        'embedding_dim':hp.IntHyperParam(min=64, max=512,default=128, step=64),
    })
}

### Create a session
session = BTBSession(tunables=tunables, scorer=scoring_function)

### Run 50 iterations for  hyperparameter tuning
best_proposal = session.run(50)

===========================CODE for HP Tuning===================================
### Questions
As highlighted in the CODE:
1. Should I use **evaluate(synthetic_data_input, real_data_global)** as the function to calculate **score** in **scoring_function()** to determine the best Hyperparameters Settings and are there any other function should I consider to calculate **score?**
2. The **sythetic_data_input** to be used as the parameter for **evaluate()**, is this **sythetic_data_input** needs to be generated using the default parameters of CopulaGAN/CTGAN  before we start tuning pipeline?

========================CODE for Initial Synthetic Data================================
model = CopulaGAN()
model.fit(real_data_global)

synthetic_data_input = model.sample(int(num_samples))
========================CODE for Initial Synthetic Data================================

### Conclusion
To conclude, the steps should be in order:
STEP_1: CODE for Initial Synthetic Data  ---> **synthetic_data_input** is generated
STEP_2: Then use the generated **synthetic_data_input** from STEP_1 to use as a parameter for  **evaluate(synthetic_data_input, real_data_global)**, given **evaluate()** is the score for the scoring_function in CODE for HP Tuning.


Is it the right approach to Hypertune the model using **synthetic_data_input** as a baseline, as this synthetic data is generated with respect to default hyperparameters of a given model?

Thanks! :)

**Comment by csala** [comment]
> * Should I use **evaluate(synthetic_data_input, real_data_global)** as the function to calculate **score** in **scoring_function()** to determine the best Hyperparameters Settings and are there any other function should I consider to calculate **score?**

Yes, the `evaluate` function is the best approach.

In any case, remember that you have the option to select which metrics are applied. So, since your dataset represents a classification task, you may want to select the ML Efficacy metrics and pass a metadata that specifies which is the `target` column.

You can see the complete list of single table metrics [here](https://sdv.dev/SDV/user_guides/evaluation/single_table_metrics.html)
And you can see the full description of the `evaluate` function [here](https://sdv.dev/SDV/api_reference/api/sdv.evaluation.evaluate.html)

> * The **sythetic_data_input** to be used as the parameter for **evaluate()**, is this **sythetic_data_input** needs to be generated using the default parameters of CopulaGAN/CTGAN  before we start tuning pipeline?

No, this is not necessary because the `BTBSession` [already does this for you](https://mlbazaar.github.io/BTB/api/btb.session.html#btb.session.BTBSession.propose): when you start a new session, the first hyperparameter proposal are always the default hyperparameters of the model, so the tuners always know what the default score is.

In other words, all you need is what you pasted above in the `CODE for HP Tuning` block.
